doc/radosgw/pools.rst

```diff
@@ -55,3 +55,54 @@ the following pool entries use namespaces of the ``us-east.rgw.meta`` pool::
     "user_uid_pool": "us-east.rgw.meta:users.uid",
 
 .. _`Multisite Configuration`: ../multisite
+
+
+.. _create-ec-pool: 
+
+Create an Erasure-coded Pool
+----------------------------
```

Comment:
there's a little too much focus on the erasure coding part here. i'd like to break this up a bit so it integrates better with the rest of this document - particularly with the `Tuning` section above:
> When ``radosgw`` first tries to operate on a zone pool that does not exist, it will create that pool with the default values from ``osd pool default pg num`` and ``osd pool default pgp num``. These defaults are sufficient for some pools, but others (especially those listed in ``placement_pools`` for the bucket index and data) will require additional tuning.

two things i'd add to that `Tuning` section:
* radosgw creates replicated pools by default 
* if you want to change any defaults, including replicated vs. erasure, you must pre-create these pools before starting radosgw

the `Associate the pool with the "rgw" application.` step is required for all rgw pools, whether replicated or erasure coded. maybe instead of `Create an Erasure-coded Pool`, this section could be called something like `Customize Pools for RGW`? it would list the same 2 steps (create pool, associate it with rgw), with some extra notes:
* erasure coding should only be used for object data pools, because most other pools require replicated
* bucket index pools should be on very fast media
* in multisite, the rgw log pool should also be on fast media

`Customize Pools for RGW` can just link to the rados docs for its steps (`createpool` and `associate-pool-to-application`) instead of spelling them out, so that this section doesn't get stale if the rados docs are changed later

to support the goal of making these instructions copy-and-pasteable, we can finish with another section `Example: Erasure Coded Pool for Object Data` that spells out the exact commands to create and associate the "default.rgw.buckets.data" pool

then the note under `adding_a_storage_class` in placement.rst could link directly to this example

what do you think  @zdover23?

Reply:
@cbodley: Give me a day, and I'll have an edit back to you. I'll need a couple of hours to digest your comments.

Reply:
@cbodley:

I intend to integrate all of the raw information in your response into the documentation, because as I read it I can feel it eradicating my ignorance and that means that it should be incorporated into the documentation that we already have. But I want to insist on keeping at least some section's name as "Creating an Erasure Coded Pool", and here's why: this PR was created in response to a complaint by someone new to the Ceph project who was searching for a section that would explain how to create an erasure coded pool, and they couldn't find such a section. I want a person in that situation to find a section that speaks directly to their concern.

I want to make it perfectly clear that I do understand that the creation of an erasure coded pool is just one instance of a pool that has been customized for RGW. But I don't think that that understanding is useful to someone new to Ceph who just wants to see erasure coding in action as quickly as possible.

In a way, this hearkens back to one of my oldest gripes about Linux-related documentation: there is a wealth of information available, but very little of it is in the form of tutorials written for the eager uninitiated.

By the time the work week begins, I'll force-push an update that should advance the conversation better than this note does.

Reply:
These two bullet points have been added to the Tuning section:


*    radosgw creates replicated pools by default
*    if you want to change any defaults, including replicated vs. erasure, you must pre-create these pools before starting radosgw

Reply:
@cbodley:

I think we have to meet the reader on the grounds of their ignorance. 

My opinion is that the information in this response is excellent, but it doesn't meet the first-time reader on the grounds of the first-time reader's ignorance.

This whole situation provides a good example of one of the fundamental tensions in the field of technical writing: one the one hand there is the correct description of something by the minds that conceived it, and on the other hand there is the mind-of-the-beginner understanding of it by people who are interested in it but don't yet understand it.

If the section is called "Customize Pools for RGW", then it won't answer the question "How do I set up an erasure coded pool?" It'll answer the question "How do I customize a pool for Rados Gateway?". But that wasn't the question that provoked this section into being. The young developer who was disappointed in the documentation didn't know that they were looking for information about how to customize a pool for RGW. The young developer was looking for information about how to set up an erasure coded pool. I'll probably be embarrassed in the future when I read that I am here invoking Gottlob Frege to make this point, but here I go, anyway: even though these two questions have the same reference (how do I customize a pool for RGW, specifically in the erasure-coded flavor?), they don't have the same sense. I want to write a section that satisfies a reader who knows only the sense of the question that was in the mind of that new developer.

**HOWEVER**
You're not wrong about the nature of pool customization and (from the perspective of RGW and pool customization) erasure coded pools just being one variety of pool customization. The understanding of this that you have articulated is (of course) the more comprehensive and correct one (as compared to the understanding of someone guided by nothing more than the wish to see an erasure coded pool in action).

So: I'd like to add your comments to the docs, but I would also like to preserve a tutorial section for people who do a Google search for "How do I create an erasure coded pool in Ceph?" I am willing to pay the cost of keeping this tutorial section up to date, for as long as the Linux Foundation sees fit to keep me around.

**SO WHAT NOW THEN**
It seems to me that (assuming that we can work this out) we should find a place for a simple "How to set up an erasure coded pool in Ceph" section that is out of the way of the more comprehensive, more correct information on this page. So: where should I put it?

Reply:
thanks @zdover23, that helps to clarify the target audience. by itself, `How do I set up an erasure coded pool?` is a ceph question, not an rgw one. we don't want to send google searches directly into RGW docs

my search for `ceph How do I set up an erasure coded pool?` came up with two hits:
* https://docs.ceph.com/en/latest/rados/operations/erasure-code/
* https://docs.ceph.com/en/latest/dev/erasure-coded-pool/

at the very top of that first doc is an example, `Creating a sample erasure coded pool`. as a whole, that doc looks pretty comprehensive - i'm not sure what more we need to add there

Reply:
any more thoughts here? does that https://docs.ceph.com/en/latest/rados/operations/erasure-code/#creating-a-sample-erasure-coded-pool section accomplish our goal, or is more detail needed?

separately, i'd love to follow up on the rgw pool docs

Reply:
> any more thoughts here? does that https://docs.ceph.com/en/latest/rados/operations/erasure-code/#creating-a-sample-erasure-coded-pool section accomplish our goal, or is more detail needed?
> 
> separately, i'd love to follow up on the rgw pool docs

@cbodley: I do have more thoughts. Or, maybe more correctly, I have thoughts similar to the ones that I've already expressed, but thoughts that can be crammed into the Procrustean bed of your questions to advance the discussion.

I don't think that the section that you link to achieves the same goal as this new procedure. Here's the problem: the section in the RADOS documentation that you've linked to is too smart. It assumes familiarity with the procedure of creating an EC pool. It's reference material for people who already have an idea of what they're doing. It's not a step-by-step hand-holding spoon-feeder for raw clueless noobs.

I don't know where to put the kind of tutorial information that I have created in this PR, but I think that it should be put somewhere. The problem (as you have identified) is that without a dedicated Guide for Ceph Noobs, information like this is going to tread on the toes of the adept and threaten to pollute their grimoires. When I meet with Josh tomorrow, I will discuss with him the creation of a Guide for the Perplexed and New.

Re: rgw pool docs: I am ready to help in any way I can.

Reply:
@cbodley Also, I'd like to make an editorial pass over https://docs.ceph.com/en/latest/rados/operations/erasure-code/#creating-a-sample-erasure-coded-pool in order to improve the flow of those sentences.

Reply:
thanks @zdover23, i'm still keen to do most of the `Customize Pools for RGW` stuff we discussed, just with a little less focus on EC specifically. if we do add a better tutorial for EC, we can link to it from there